src/server/mod.rs

```diff
@@ -1,5 +1,387 @@
-use super::super::common::{screen};
+use crate::client::ClientInstruction;
+use crate::common::{ChannelWithContext, SenderType, SenderWithContext};
+use crate::errors::{ContextType, ErrorContext, OsContext, PtyContext, ServerContext};
+use crate::os_input_output::{ServerOsApi, ServerOsApiInstruction};
+use crate::panes::PaneId;
+use crate::pty_bus::{PtyBus, PtyInstruction};
+use crate::screen::ScreenInstruction;
+use crate::{cli::CliArgs, common::pty_bus::VteEvent};
+use serde::{Deserialize, Serialize};
+use std::os::unix::io::RawFd;
+use std::path::PathBuf;
+use std::sync::mpsc::channel;
+use std::thread;
+use zellij_tile::prelude::InputMode;
 
-pub fn start_server() {
-  // TODO
-}
\ No newline at end of file
+/// Instructions related to server-side application including the
+/// ones sent by client to server
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum ServerInstruction {
+    OpenFile(PathBuf),
+    SplitHorizontally,
+    SplitVertically,
+    MoveFocus,
+    NewClient(String),
+    ToPty(PtyInstruction),
+    ToScreen(ScreenInstruction),
+    OsApi(ServerOsApiInstruction),
+    DoneClosingPane,
+    ClosePluginPane(u32),
+    ClientExit,
+    Exit,
+}
+impl ServerInstruction {
+    // ToPty
+    pub fn spawn_terminal(path: Option<PathBuf>) -> Self {
+        Self::ToPty(PtyInstruction::SpawnTerminal(path))
+    }
```

Comment:
@khs26, as discussed in the last meeting, this is the `ServerInstruction` enum which the client uses to send messages to the server. The nested enums (`PtyInstruction` and `ScreenInstruction`) made the `send()` method calls too verbose, so we have defined these helper methods for now.

We have an option to flatten the `ServerInstruction` enum so every `ScreenInstruction` and `PtyInstruction` would be directly defined used the `ScreenInstruction` enum as well. IMO this will be a better option once we have macros for generating contexts from the Instructions and the screen and plugin threads are moved to the server-side (I'm working on this right now).

So if you have any other ideas, let us know.

Reply:
Oh, and maybe also implement `From<ServerInstruction>` and `From<PtyInstruction>` for `ServerInstruction` actually does the job when we're having to manipulate these before serializing/sending. :) I should have maybe thought of that first...

Reply:
I pushed a change with that in - feel free to discard it if that doesn't help in any way

Reply:
Actually, reading the thread a bit more, I think I got that the problem is just that we don't want to have `ScreenInstruction` and `PtyInstruction` and friends on the client side, is that right? (also, oops...)

If so, I think the main issue we have to solve is defining what we want the client-side instructions to be.

We'd then similarly serialize those messages and not worry about what the server has to do on its side to handle them. My instinct would be to have `input_handler` on the client side just serialize `Action` and ship it over to the server. @imsnif, do you think that makes sense?

The implication of all this would be that the client side "owns":
  - Hotkey definitions
  - Input mode
  - A copy of whatever data they need from the server side to render a screen